src/ring_theory/discriminant.lean

```diff
@@ -0,0 +1,222 @@
+
+open matrix finite_dimensional fintype polynomial finset
+
+namespace algebra
+
+variables (A : Type u) {B : Type v} (C : Type z) {ι : Type w}
+variables [comm_ring A] [comm_ring B] [algebra A B] [comm_ring C] [algebra A C]
+
+section matrix
+
+/-- Given an `A`-algebra `B` and `b`, an `ι`-indexed family of elements of `B`, we define
+`trace_matrix A ι b` as the matrix whose `(i j)`-th element is the trace of `b i * b j`. -/
+@[simp] noncomputable
+def trace_matrix (b : ι → B) : matrix ι ι A
+| i j := trace_form A B (b i) (b j)
+
+lemma trace_matrix_def (b : ι → B) : trace_matrix A b = λ i j, trace_form A B (b i) (b j) := rfl
+
+variable {A}
+
+lemma trace_matrix_of_matrix_vec_mul [fintype ι] (b : ι → B) (P : matrix ι ι A) :
+  trace_matrix A ((P.map (algebra_map A B)).vec_mul b) = Pᵀ ⬝ (trace_matrix A b) ⬝ P :=
+begin
+  ext α β,
+  rw [trace_matrix, vec_mul, dot_product, vec_mul, dot_product, mul_apply, bilin_form.sum_left,
+    fintype.sum_congr _ _ (λ (i : ι), @bilin_form.sum_right _ _ _ _ _ _ _ _
+    (b i * P.map (algebra_map A B) i α) (λ (y : ι), b y * P.map (algebra_map A B) y β)), sum_comm],
+  congr, ext x,
+  rw [mul_apply, sum_mul],
+  congr, ext y,
+  rw [map_apply, trace_form_apply, mul_comm (b y), ← smul_def],
+  simp only [id.smul_eq_mul, ring_hom.id_apply, map_apply, transpose_apply, linear_map.map_smulₛₗ,
+    trace_form_apply, algebra.smul_mul_assoc],
+  rw [mul_comm (b x), ← smul_def],
+  ring_nf,
+  simp,
+end
+
+lemma trace_matrix_of_matrix_mul_vec [fintype ι] (b : ι → B) (P : matrix ι ι A) :
+  trace_matrix A ((P.map (algebra_map A B)).mul_vec b) = P ⬝ (trace_matrix A b) ⬝ Pᵀ :=
+begin
+  refine add_equiv.injective transpose_add_equiv _,
+  rw [transpose_add_equiv_apply, transpose_add_equiv_apply, ← vec_mul_transpose,
+    ← transpose_map, trace_matrix_of_matrix_vec_mul, transpose_transpose, transpose_mul,
+    transpose_transpose, transpose_mul]
+end
+
+lemma trace_matrix_of_basis [decidable_eq ι] [fintype ι] (b : basis ι A B) :
+  trace_matrix A b = bilin_form.to_matrix b (trace_form A B) :=
+begin
+  ext i j,
+  rw [trace_matrix, trace_form_apply, trace_form_to_matrix]
+end
+
+variable (A)
+
+/-- `embeddings_matrix A C b : matrix ι (B →ₐ[A] C) C` is the matrix whose `(i, σ)` coefficient is
+  `σ (b i)`. It is mostly useful for fields when `fintype.card ι = finrank A B` and `C` is
+  algebraically closed. -/
+@[simp] def embeddings_matrix (b : ι → B) : matrix ι (B →ₐ[A] C) C
+| i σ := σ (b i)
+
+/-- `embeddings_matrix_reindex A C b e : matrix ι ι C` is the matrix whose `(i, j)` coefficient
+  is `σⱼ (b i)`, where `σⱼ : B →ₐ[A] C` is the embedding corresponding to `j : ι` given by a
+  bijection `e : ι ≃ (B →ₐ[A] C)`. It is mostly useful for fields and `C` is algebraically closed.
+  In this case, in presnce of `h : fintype.card ι = finrank A B`, one can take
+  `e := equiv_of_card_eq ((alg_hom.card A B C).trans h.symm)`. -/
+def embeddings_matrix_reindex (b : ι → B) (e : ι ≃ (B →ₐ[A] C)) :=
```

Comment:
I recall defining this matrix (and the previous ones) for the proof of [`det_trace_form_ne_zero`](https://github.com/leanprover-community/mathlib/blob/master/src/ring_theory/trace.lean#L348), but ended up inlining it apparently. It's probably worth the time to move `embeddings_matrix_reindex` and some of its lemmas to `trace.lean` and seeing if we can use them to simplify the proofs there.

If you don't have the time for that right now, I'll see if I can do that later today.

Reply:
I don't have time today to do it, but if you are busy feel free to suggest here what you think it should be done and I will take care of it.

Reply:
I'm happy to do it now, while the code is still fresh in my mind. Do you mind if I push directly to your branch?

Reply:
No problem!

Reply:
Done!